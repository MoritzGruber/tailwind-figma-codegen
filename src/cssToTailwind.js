// this is in js bcz `babel-loader` doesn't support ts

import { TailwindConverter } from 'css-to-tailwindcss';

const converter = new TailwindConverter({
	remInPx: 16,
	arbitraryPropertiesIsEnabled: true
});

/**
 *
 * @param cssObj {Record<string, string>}
 * @returns
 */
export const cssToTailwind = async (cssObj) => {
	if (cssObj['font-style'] === 'normal') {
		delete cssObj['font-style'];
	}

	const css = Object.entries(cssObj)
		.map(([key, value]) => `${key}: ${value.replace(/\/\*.*\*\//g, '').trim()};`)
		.join('\n');

	const { convertedRoot, nodes } = await converter.convertCSS(`
    div {
        ${css}
    }
    `);

	const doc = {
		className: nodes[0].tailwindClasses.join(' ').trim(),
		css: convertedRoot.toString()
	};

	return doc;
};
