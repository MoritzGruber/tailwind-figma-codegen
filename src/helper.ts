
export function rgbToHex(int: any) {
    let hex = Number(int).toString(16);
    if (hex.length < 2) {
      hex = `0${hex}`;
    }
    return hex;
  }
  
  export function makeHex(r:any, g:any, b:any, a:any) {
    const red = rgbToHex(r);
    const green = rgbToHex(g);
    const blue = rgbToHex(b);

    if (a !== undefined) {
      const alpha = rgbToHex(Math.round(a * 255));
      return `#${red}${green}${blue}${alpha}`.toUpperCase();
    }

    return `#${red}${green}${blue}`.toUpperCase();
  }
  
  export function makeRgb(color: any) {
    const r = Math.round(255 * color.r);
    const g = Math.round(255 * color.g);
    const b = Math.round(255 * color.b);
    // TODO * 100 and round
    const a = Math.round(100 * color.a) / 100;
    return { r, g, b, a };
  }