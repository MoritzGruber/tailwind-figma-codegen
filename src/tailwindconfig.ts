export const tailwindifyName = (str: string) => {
    return str
    .replace(/\//g, '-')  // Replace all '/' with '-'
    .replace(/\s/g, '-')  // Replace all spaces with '-'
    .replace(/--+/g, '-');  // Replace two or more '-' with a single '-'};
};

export const variablesToTailwindConfig = async (variables: Record<string, any>) => {

    const colors:any = {}
    
    variables.colors.map(
        (c: any) => {
            colors[tailwindifyName(c.name)] = c.value;
        } 
    );

	const tailwindConfig = {
		theme: {
			extend: {
				colors: {
					...colors
				},
			}
		}
	};

	return tailwindConfig;
};
