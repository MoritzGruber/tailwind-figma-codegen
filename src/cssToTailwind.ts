import { TailwindConverter } from 'css-to-tailwindcss';

import type { Config } from 'tailwindcss';
import { makeHex } from './helper';

function replaceCssVarWithFallback(cssString: string) {
	return cssString.replace(/var\((.*?),\s*(.*?)\)/g, (_, varName: string, fallback) => {
		return fallback.trim();
	});
}

export function replaceRgbaWithHex(cssString: string): string {
	const rgbaRegex = /rgba\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*,\s*([\d.]+)\s*\)/g;

	return cssString.replace(rgbaRegex, (match, r, g, b, a) => {
		const hexColor = makeHex(parseInt(r, 10), parseInt(g, 10), parseInt(b, 10), parseFloat(a));
		return hexColor;
	});
}

export const postProcessTailwindClasses = (classes: string) => {
	const postReplacements = [
		['leading-[normal]', 'leading-normal'],
		['leading-[1.5]', 'leading-normal'],
		['leading-[150%]', 'leading-normal'],
		['leading-[loose]', 'leading-loose'],
		['leading-[none]', 'leading-none'],
		['leading-[tight]', 'leading-tight'],
		['leading-[snug]', 'leading-snug'],
		['leading-[relaxed]', 'leading-relaxed'],
		['tracking-[0px]', 'tracking-normal'],
	];

	return postReplacements.reduce((acc, [search, replace]) => acc.replace(search, replace), classes);
};

export const cssToTailwind = async (cssObj: Record<string, string>, ignoreFields?: string[], tailwindConfig?: undefined | Config) => {
	const converter = new TailwindConverter(
		tailwindConfig
			? {
					remInPx: 16,
					arbitraryPropertiesIsEnabled: false,
					tailwindConfig
				}
			: { remInPx: 16, arbitraryPropertiesIsEnabled: false }
	);

	if (ignoreFields && ignoreFields.length) {
		for (const field of ignoreFields) {
			if (field.includes('=')) {
				const [key, value] = field.split('=');
				if (value === '*' && cssObj[key]) {
					delete cssObj[key];
				}

				if (cssObj[key] === value) {
					delete cssObj[key];
				}
			} else delete cssObj[field];
		}
	}

	const css = Object.entries(cssObj)
		.map(([key, value]) => {
			if (key == 'background') {
				// for just colors we need to change this to background color
				const twColorNames: any = [];
				if (tailwindConfig && tailwindConfig.theme) {
					tailwindConfig.theme.colors &&
						Object.keys(tailwindConfig.theme.colors).forEach((colorName) => {
							twColorNames.push(colorName);
						});
					if (tailwindConfig.theme.extend) {
						tailwindConfig.theme.extend.colors &&
							Object.keys(tailwindConfig.theme.extend.colors).forEach((colorName) => {
								twColorNames.push(colorName);
							});
					}
				}

				if (
					value.includes('rgb') ||
					value.includes('rgba') ||
					value.includes('#') ||
					value.includes('hsl') ||
					value.includes('hsla') ||
					twColorNames.some((colorName: any) => value.includes(colorName))
				) {
					key = 'background-color';
				}
			}

			return `${key}: ${value.replace(/\/\*.*\*\//g, '').trim()};`;
		})
		.join('\n');

	console.log(` css`, css);
	const replacedCss = replaceRgbaWithHex(replaceCssVarWithFallback(css));
	console.log(` replacedCss`, replacedCss);

	const { convertedRoot, nodes } = await converter.convertCSS(`
    div {
        ${replacedCss}
    }
    `);
	console.log(` convertedRoot.toString()`, convertedRoot.toString());

	const doc = {
		className: postProcessTailwindClasses(nodes[0].tailwindClasses.join(' ').trim()),
		css: convertedRoot.toString()
	};

	return doc;
};
