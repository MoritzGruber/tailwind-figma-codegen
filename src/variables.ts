import { makeHex, makeRgb } from './helper';

export async function loadTextStyles() {
	// eslint-disable-next-line
	const textStyles = await figma.getLocalTextStylesAsync();
	console.log(`figma textStyles`, textStyles);
	const fontSizes: any[] = [];
	const fontFamilies: any[] = [];
	const finalSizes: any[] = [];
	const finalFamilies: any[] = [];

	textStyles.forEach((style) => {
		const { family } = style.fontName;
		const { fontSize } = style;

		fontFamilies.push(family);
		fontSizes.push(fontSize);
	});

	// Get unique values
	const singleSizes = Array.from(new Set(fontSizes)).sort((a, b) => a - b);
	const singleFamilies = Array.from(new Set(fontFamilies));

	// Clean sizes
	singleSizes.forEach((size) => {
		const name = '';
		// Pass everything as a string
		const value = size.toString();
		const result = { name, value };
		finalSizes.push(result);
	});

	// Clean families
	singleFamilies.forEach((family) => {
		const name = family.replace(/\s+/g, '-').toLowerCase();
		const value = family;
		const result = { name, value };
		finalFamilies.push(result);
	});

	// Make objects
	return { fontSizes: finalSizes, fontFamilies: finalFamilies };
}

export const loadColors = async () => {
	// eslint-disable-next-line
	const colorStyles = await figma.getLocalPaintStylesAsync();
	const colors: any[] = []; // array of hex values and their names
	const gradientColors: any[] = [];

	colorStyles.forEach((style) => {
		// Extra check for empty paint styles
		const paint: any = style.paints[0] || null;
		if (paint) {
			const { color = null, gradientStops = null, opacity = null } = style.paints[0] as any;
			/* Only work with solid colors */
			if (color) {
				const { name } = style;
				if (name.includes('primary')) {
					console.log(` color ${name}`, { color, style, opacity });
				}
				let alpha = undefined;
				if (opacity != 1 && isNaN(opacity) === false) {
					alpha = Math.round(100 * opacity) / 100;
				} 
				const { r, g, b } = makeRgb(color);
					const value = makeHex(r, g, b, alpha);
					const result = { name, value };
					colors.push(result);
			} else if (gradientStops && gradientStops.length > 0) {
				/* Add gradients as a suggestion */
				gradientStops.forEach((stop: any) => {
					const { r, g, b } = makeRgb(stop.color);
					const value = makeHex(r, g, b, undefined);
					gradientColors.push(value);
				});
			}
		}
	});
	return { colors, gradientColors };
};

export async function loadNodeStyles() {
	// eslint-disable-next-line no-undef
	await figma.loadAllPagesAsync();
	// eslint-disable-next-line no-undef
	const filterdNodes = figma.root.findAll((n: any) => n.cornerRadius);
	const radii = new Set();
	Array.from(filterdNodes).forEach((n: any) => {
		if (typeof n.cornerRadius === 'number') {
			const value = n.cornerRadius < 99 ? n.cornerRadius : 999;
			radii.add(value);
		}
	});
	const radiiArray = [...radii].sort((a, b) => a - b);
	const finalRadii = [];

	// const position = radiiArray.indexOf(16);
	radiiArray.forEach((radius) => {
		// const n = calculatePosition(i, position, radiiArray.length);
		// Rename base to default
		// const value = radius > 98 ? '9999px' : `${radius / 16}rem`;
		const value = Number(radius);
		const name = '';
		finalRadii.push({ name, value });
	});

	// Add default none
	finalRadii.unshift({ name: 'none', value: 0 });
	return { radius: finalRadii };
}

export async function loadEffectStyles() {
	// eslint-disable-next-line
	const effectStyles = await figma.getLocalEffectStylesAsync();
	const shadows: any[] = [];

	effectStyles.forEach((style) => {
		const shadowStyle: any = {};
		const { effects, name } = style;
		const styleString: any[] = [];
		// Generate css string for each shadow (if it has a color, BACKGROUND_BLUR does not)
		effects.forEach((effect) => {
			const { color, offset, radius, spread, type } = effect as any;
			if (color) {
				const { r, g, b, a } = makeRgb(color);
				const colorString = `${r},${g},${b},${a}`;
				styleString.push(
					`${type === 'INNER_SHADOW' ? 'inset ' : ''}${offset.x}px ${offset.y}px ${radius}px ${spread}px rgba(${colorString})`
				);
			}
		});
		// Create object & push it to shadows
		if (shadowStyle) {
			shadowStyle.name = name;
			shadowStyle.value = styleString.join(', ');
			shadows.push(shadowStyle);
		}
	});

	return { shadows };
}

export const laodAllVariables = async () => {
	const colorPromise = loadColors();
	const textPromise = loadTextStyles();
	const nodePromise = loadNodeStyles();
	const effectPromise = loadEffectStyles();

	const variables = {
		__loaded: true,
		...(await colorPromise),
		...(await textPromise),
		...(await nodePromise),
		...(await effectPromise)
	};
	return variables;
};
