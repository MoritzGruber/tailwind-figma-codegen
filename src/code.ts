import { cssToTailwind } from './cssToTailwind';
import { variablesToTailwindConfig } from './tailwindconfig';
import { loadEffectStyles, loadNodeStyles, loadColors, loadTextStyles, laodAllVariables } from './variables';

const defaultState = {
	ignoreFields: 'font-style=normal',
	tailwindConfig: ''
};

let state = defaultState;

let loaded = false;
// Save data
// await figma.clientStorage.setAsync('user-preferences', {theme: 'dark'})

// Retrieve data
// const preferences = await figma.clientStorage.getAsync('user-preferences')

const loadData = async () => {
	const variables = {
		colors: await loadColors(),
		textStyles: await loadTextStyles(),
		nodeStyles: await loadNodeStyles(),
		effectStyles: await loadEffectStyles()
	};
	console.log(`variables `, variables);
	await figma.clientStorage
		.getAsync('state')
		.then((data) => {
			console.log('Loaded user data:', data);
			state = data;
			loaded = true;
			// Use the data to initialize your plugin state
		})
		.catch((error) => {
			console.error('Error loading user data:', error);
		});
};
loadData();

figma.codegen.on('generate', async (e) => {
	const node = e.node;
	if (!loaded) {
		await loadData();
	}
	const cssObj = await node.getCSSAsync();

	console.log(`variables `, await laodAllVariables());
	console.log(` cssObj`, cssObj);
	let configObj = undefined;
	try {
		if (state.tailwindConfig && state.tailwindConfig.trim() != '') {
			configObj = JSON.parse(state.tailwindConfig);
		}
	} catch (e) {
		console.error('Invalid tailwindConfig', state && state.tailwindConfig);
	}
	const ignoreFields = state ? state.ignoreFields : defaultState.ignoreFields;

	const twConfig: any = await variablesToTailwindConfig(await laodAllVariables());
	console.log(` twConfig`, twConfig);

	const { className, css } = await cssToTailwind(
		cssObj,
		ignoreFields.split(',').map((field) => field.trim()),
		twConfig
	);

	return [
		{
			title: 'tailwindcss',
			code: className,
			language: 'CSS'
		},
		{
			title: 'css',
			code: css,
			language: 'CSS'
		}
	];
});

figma.codegen.on('preferenceschange', async (e) => {
	if (e.propertyName == 'ignoreFields') {
		figma.showUI(
			`
			<div>
				<h1>Ignore Fields</h1>
				<p>Enter the fields you want to ignore separated by comma. You can add a value using <kbd>key=value</kbd> format to ignore a field only when this field matches the value</p>
				<input type="text" value="${state.ignoreFields}" id="ignoreFields" />
				<button id="save" onclick="parent.postMessage({ pluginMessage: { type: 'save', ignoreFields: document.getElementById('ignoreFields').value } }, '*')">Save</button>
			</div>
			`,
			{ visible: true }
		);
	}
	if (e.propertyName == 'tailwindConfig') {
		const twConfig: any = await variablesToTailwindConfig(await laodAllVariables());

		figma.showUI(
			`
     <style>
        body { font-family: Arial, sans-serif; padding: 1rem; }
        .container { max-width: 28rem; margin: 0 auto; }
        h1 { font-size: 1.5rem; font-weight: bold; margin-bottom: 1rem; }
        textarea { width: 100%; height: 12rem; padding: 0.5rem; border: 1px solid #ccc; border-radius: 0.25rem; margin-bottom: 0.5rem; }
        #validationMessage { font-size: 0.875rem; margin-bottom: 0.5rem; }
        button { background-color: #3b82f6; color: white; font-weight: bold; padding: 0.5rem 1rem; border: none; border-radius: 0.25rem; cursor: pointer; }
        button:hover { background-color: #2563eb; }
        button:disabled { background-color: #9ca3af; cursor: not-allowed; }
        .text-green-600 { color: #059669; }
        .text-red-600 { color: #dc2626; }
    </style>
    <div class="container">
        <h1>Tailwind Config</h1>
        <textarea id="tailwindConfig" value="${JSON.stringify(twConfig)}"></textarea>
    </div>
			`,
			{ visible: true }
		);
	}
});

figma.ui.onmessage = (message) => {
	console.log(` onmessage`);
	if (message.type === 'save') {
		if (!state) {
			state = defaultState;
		}
		if (message.ignoreFields) {
			state.ignoreFields = message.ignoreFields;
		}
		if (message.tailwindConfig) {
			state.tailwindConfig = message.tailwindConfig;
		}
		figma.clientStorage.setAsync('user-preferences', { theme: 'dark' }).then(
			() => {
				figma.notify('Data updated');
				figma.ui.close();
				figma.codegen.refresh();
			},
			(error) => {
				console.error('Error saving user data:', error);
				figma.notify('Error saving data');
				figma.ui.close();
				figma.codegen.refresh();
			}
		);
	}
};
