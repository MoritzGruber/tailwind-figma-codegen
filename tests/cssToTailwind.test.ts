import { describe, expect, test } from 'vitest';
import { cssToTailwind } from '../src/cssToTailwind';

describe('cssToTailwind', () => {
	test('should convert css to tailwind', () => {
		expect(
			cssToTailwind({
				display: 'flex',
				width: '404px',
				height: '336px',
				padding: '68px 16px 16px 16px',
				'flex-direction': 'column',
				'align-items': 'flex-start',
				gap: '64px',
				'flex-shrink': '0'
			})
		).resolves.toMatchSnapshot();
	});

	test('should remove font-style: normal', () => {
		expect(
			cssToTailwind({
				color: 'var(--Primary-A-primary-a-400, #464646)',
				'font-family': 'Montserrat',
				'font-size': '16px',
				'font-style': 'normal',
				'font-weight': '500',
				'line-height': '20px'
			})
		).resolves.toMatchSnapshot();
	});

	test('should remove inline comments', () => {
		expect(
			cssToTailwind({
				color: 'var(--Primary-A-primary-a-400, #464646)',
				'font-family': 'Montserrat',
				'font-size': '16px',
				'font-style': 'normal',
				'font-weight': '500',
				'line-height': '20px /* 125% */'
			})
		).resolves.toMatchSnapshot();
	});

	test('should add bg-color class', () => {
		expect(
			cssToTailwind({
				'border-radius': '12px',
				background: 'var(--Primary-B-primary-b-600, #D1D1D1)'
			})
		).resolves.toMatchSnapshot();
	});
});
