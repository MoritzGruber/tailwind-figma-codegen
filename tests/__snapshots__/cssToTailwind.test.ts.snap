// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cssToTailwind > should add bg-color class 1`] = `
{
  "className": "bg-[#D1D1D1] rounded-xl",
  "css": "div {
    @apply bg-[#D1D1D1] rounded-xl;
}
    ",
}
`;

exports[`cssToTailwind > should convert css to tailwind 1`] = `
{
  "className": "flex w-[404px] h-[336px] flex-col items-start gap-16 shrink-0 pt-[68px] pb-4 px-4",
  "css": "div {
    @apply flex w-[404px] h-[336px] flex-col items-start gap-16 shrink-0 pt-[68px] pb-4 px-4;
}
    ",
}
`;

exports[`cssToTailwind > should remove font-style: normal 1`] = `
{
  "className": "text-[#464646] text-base not-italic font-medium leading-5",
  "css": "div {
    @apply text-[#464646] text-base not-italic font-medium leading-5;
    font-family: Montserrat;
}
    ",
}
`;

exports[`cssToTailwind > should remove inline comments 1`] = `
{
  "className": "text-[#464646] text-base not-italic font-medium leading-5",
  "css": "div {
    @apply text-[#464646] text-base not-italic font-medium leading-5;
    font-family: Montserrat;
}
    ",
}
`;
